tests/test_responsetypes.py

```diff
@@ -72,17 +72,35 @@ def test_from_headers(self):
             retcls = responsetypes.from_headers(source)
             assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
-    def test_from_args(self):
-        # TODO: add more tests that check precedence between the different arguments
+    def test_from_args_pre_xtractmime(self):
         mappings = [
             ({'url': 'http://www.example.com/data.csv'}, TextResponse),
-            # headers takes precedence over url
             ({'headers': Headers({'Content-Type': ['text/html; charset=utf-8']}),
-              'url': 'http://www.example.com/item/'}, HtmlResponse),
+              'url': 'http://www.example.com/item/'}, HtmlResponse), # Failing with xtractmime, returning TextResponse expected HtmlResponse
             ({'headers': Headers({'Content-Disposition': ['attachment; filename="data.xml.gz"']}),
-              'url': 'http://www.example.com/page/'}, Response),
-
+              'url': 'http://www.example.com/page/'}, Response), # Failing with xtractmime, returning TextResponse expected Response
+        ]
+        for source, cls in mappings:
+            retcls = responsetypes.from_args(**source)
+            assert retcls is cls, f"{source} ==> {retcls} != {cls}"
 
+    def test_from_args_post_xtractmime(self):
+        mappings = [
+            ({'body': b'Some plain text data with tabs and null bytes',
+              'headers': Headers({'Content-Type': ['text/html; charset=utf-8'], })},
+             TextResponse),
+            ({'body': b'\x03\x02\xdf\xdd\x23', 'headers': Headers({'Content-Encoding': 'UTF-8'})}, 
+             Response),
+            # different behaviour with http and non-http urls
+            ({'body': b'\x00\xfe\xff', 'url': 'http://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, Response), 
+            ({'body': b'\x00\xfe\xff', 'url': '://www.example.com/item/',
+              'headers': Headers({'Content-Type': b'text/plain'})}, TextResponse),
+            ({'url': 'http://www.example.com/item/file.html'}, HtmlResponse), # Failing with xtractmime, return TextResponse expected HtmlResponse
+            ({'body': b'Some plain text\ndata with tabs\t and null bytes\0'}, Response), # earlier expected to be binary response, refer "test_from_body()"
```

Comment:
I’m not sure how to best deal with this case. It’s clear that [we do not want NUL bytes to cause a body to be interpreted as binary](https://github.com/scrapy/parsel/issues/123).

One possible approach, without altering xtractmime, would be to remove NUL bytes from a body if those are the only binary data bytes in the body. I’m thinking this may be best, to keep xtractmime true to the standard, even though string-replacement in a whole response could be inefficient; maybe we could expose xtractmime’s header lenght, and only replace in that part of the body (and maybe pass only that part to xtractmime, while at it).

Another approach would be to modify xtractmime, adding an extra parameter that allows customizing what is to be understood as a binary data byte. But I think this would increase the complexity of the xtractmime implementation for no good reason.

Reply:
For this I think we can import `BINARY_BYTES` from `xtractmime._patterns.py` and remove NULL bytes from the body till it's resource header length i.e body[:1445]  :
```
for byte in BINARY_BYTES:
     body = body[:1445].replace(byte, b"") + body[1445:]
```

Reply:
We only want to remove `NUL` bytes, not all binary bytes. That is, if the body contains any other binary byte, it is OK for xtractmime to consider it binary.

What we don’t want is to remove NUL bytes _if_ the header contains any other binary bytes. We only want to remove NUL bytes if all the bytes (in the header subset) are either non-binary or NUL.

Also, we need to be careful not to leave any NUL byte unchanged in the header part. If we use `body = body[:1445].replace(byte, b"") + body[1445:]` and there is at least 1 NUL byte in the input `body[:1445]` and also one at `body[1445]`, then we would only be removing the first one and moving the other one to `body[1444]` in the outgoing `body`, not fixing the issue for such a corner case. Since the header length is rather arbitrary, the best approach may be not to append the remainder of the body, and instead just do `body = body[:1445].replace(byte, b"")`. The body we would be passing would be shorter than the header limit, but I cannot think of a reason why that would be a significant problem, the xtractmime algorithm would still remain “deterministic for the majority of cases”, as the standard says.

I would also suggest not to hardcode 1445 in Scrapy, and instead import the number from xtractmime, where it could be exposed as a public constant (`RESOURCE_HEADER_BUFFER_LENGTH`? `RESOURCE_HEADER_MAX_LENGTH`?).